Atmos init speedup, saves 4 seconds

LemonInTheDark opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
Micro optimizes ssair's turf init, saving 2 seconds

Most of this is making existing operations do more legwork, or cheaper. I did add cycle checking to ONLY init turf linking, which required creating a new proc.
Did some horrible horrible things in said proc to save like 0.8 seconds. I think it was worth it.


Hey components please stop doing 4 lowertext() calls to check 1 pair  of atmos component neighbors


Moves gas parsing onto ssair, caches parsing

This saves like 1.4 seconds just because we don't need to walk gas ids
for the 1000kth time

I moved it onto the turf/ssair namespace because it lets us do copy(),
which is cheaper then copy_thing or whatever it's called

## Why It's Good For The Game

Speed

LemonInTheDark committed:
Micro optimizes ssair's turf init, saving 2 seconds

Most of this is making existing operations do more legwork, or cheaper.
I did add cycle checking to ONLY init turf linking, which required
creating a new proc.
Did some horrible horrible things in said proc to save like 0.8 seconds.
I think it was worth it.

LemonInTheDark committed:
Hey components please stop doing 4 lowertext() calls to check 1 pair of atmos component neighbords

Kapu1178 commented:
you're welcome for telling you how bad `get_step_multiz()` is.

Kylerace commented on code/controllers/subsystem/air.dm:
```diff
@@ -521,7 +521,7 @@ SUBSYSTEM_DEF(air)
 /datum/controller/subsystem/air/proc/setup_allturfs()
 	var/list/turfs_to_init = block(locate(1, 1, 1), locate(world.maxx, world.maxy, world.maxz))
 	var/list/active_turfs = src.active_turfs
-	var/times_fired = ++src.times_fired
```
muh cache

LemonInTheDark replied:
it's only used once, since I replaced it with time.
tho now that I mention it. I need to figure out how to ensure time doesn't block valid share attempts

Kylerace commented on code/modules/atmospherics/environmental/LINDA_system.dm:
```diff
@@ -53,27 +53,79 @@
+	// You know it's gonna be stupid when they include a unit test in the atmos code
+	// Yes, inlining the string concat does save 0.1 seconds
+	#ifdef UNIT_TESTS
+	ASSERT(UP == 16)
```
why would this change from an engine defined constant?

LemonInTheDark replied:
I do not trust lummox

Mothblocks replied:
i asked them to do this, its not so much because its expected to change but to help explain why the code is the way it is

Kylerace commented on code/modules/atmospherics/environmental/LINDA_system.dm:
```diff
@@ -53,27 +53,79 @@
+		var/turf/current_turf = (direction & (UP|DOWN)) ? \
+			(direction & UP) ? \
+				(z_traits["16"]) ? \
+					(get_step(locate(x, y, z + z_traits["16"]), NONE)) : \
```
no need for the second z_traits read here since its always going to be TRUE if it exists

Kylerace replied:
also why the get_turf() around locate()?

LemonInTheDark replied:
> no need for the second z_traits read here since its always going to be TRUE if it exists

It isn't. trait["UP"] is the AMOUNT to move up or down to land on the z level, not just IF the z level exists or not
WAIT A MOMENT IT IS WHAT THE FUCK WHY DOES THAT CODE EVEN EXIST IF IT'S NOT GONNA FUCKING WORK

KEVINNNNNNNNNNNNNNNNNNNNNN

Kylerace commented on code/modules/atmospherics/environmental/LINDA_system.dm:
```diff
@@ -53,27 +53,79 @@
+					(get_step(locate(x, y, z + z_traits["16"]), NONE)) : \
+				(null) : \
+				(z_traits["32"]) ? \
+					(get_step(locate(x, y, z + z_traits["32"]), NONE)) : \
```
this needs to be - not +, since the values of this list are always either null or +1

LemonInTheDark replied:
I thought it was a fucking offset KEVINNNNNNNN

Kylerace commented on code/modules/atmospherics/environmental/LINDA_system.dm:
```diff
@@ -53,27 +53,79 @@
+		if(canpass && CANATMOSPASS(current_turf, src, (direction & (UP|DOWN))) && !(blocks_air || current_turf.blocks_air))
+			LAZYINITLIST(current_turf.atmos_adjacent_turfs)
+			atmos_adjacent_turfs[current_turf] = TRUE
+			current_turf.atmos_adjacent_turfs[src] = TRUE
```
i wonder if it would be better to make this a non associative list since it cant exceed 6 elements

LemonInTheDark replied:
the add is somewhat trivial, tho you could test it. would save a good bit on memory.

LemonInTheDark commented:
> you're welcome for telling you how bad `get_step_multiz()` is.

You can't take credit for my sins, I wrote that affront to god before you mentioned it

LemonInTheDark commented:
I also need to check if my cycle caching effects supercon

LemonInTheDark committed:
Moves gas parsing onto ssair, caches parsing

This saves like 1.4 seconds just because we don't need to walk gas ids
for the 1000kth time

I moved it onto the turf/ssair namespace because it lets us do copy(),
which is cheaper then copy_thing or whatever it's called

LemonInTheDark committed:
kyler review
